src/algebra/module/torsion.lean

```diff
@@ -0,0 +1,113 @@
+/-
+Copyright (c) 2022 Pierre-Alexandre Bazin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pierre-Alexandre Bazin
+-/
+import algebra.module
+import linear_algebra.quotient
+import ring_theory.ideal.quotient
+
+/-!
+# Torsion submodules
+
+## Main definitions
+
+* `torsion R M a` : the `a`-torsion submodule, containing all elements `x` of `M` such that
+  `a • x = 0`.
+* `torsion' R M` : the torsion submoule, containing all elements `x` of `M` such that  `a • x = 0`
+  for some non-zero `a` in `R`.
+
+## Main statements
+
+* `torsion' R M` is a submodule when `R` is a domain.
+* `torsion R M a` can be viewed as a `(R ⧸ R·a)`-module.
+* `quot_torsion_is_torsion_free` : `(M ⧸ torsion' R M)` is a torsion-free module, that is, there is
+  no non-zero `a`, `x` such that `a • x = 0`.
+
+## Notation
+
+* The notions are defined for a `comm_ring R` and a `module R M`. We further require `R` to be a
+  domain when talking about `torsion' R M` (otherwise it may not be a submodule), and for `M` to be
+  an `add_comm_group` when it's needed.
+* The letters `a`, `b`, ... are used for scalars (in `R`), while `x`, `y`, ... are used for vectors
+  (in `M`).
+
+## Tags
+
+Torsion, submodule, module, quotient
+-/
+
+
+section defs
+variables (R M : Type*) [comm_ring R] [add_comm_monoid M] [module R M] (a : R)
```

Comment:
What happens if you relax `comm_ring` to `comm_semiring`, and replace `is_domain R` with the weaker `no_zero_divisors R` `nontrivial R` below? Do all the proofs still go through?